packages/kit/src/core/build/build_service_worker.js

```diff
@@ -0,0 +1,102 @@
+import fs from 'fs';
+import path from 'path';
+import vite from 'vite';
+import { s } from '../../utils/misc.js';
+import { deep_merge } from '../../utils/object.js';
+import { print_config_conflicts } from '../config/index.js';
+
+/**
+ * @param {{
+ *   cwd: string;
+ *   assets_base: string;
+ *   config: import('types/config').ValidatedConfig
+ *   manifest_data: import('types/internal').ManifestData
+ *   build_dir: string;
+ *   output_dir: string;
+ *   client_entry_file: string;
+ *   service_worker_entry_file: string | null;
+ * }} options
+ * @param {import('vite').Manifest} client_manifest
+ */
+export async function build_service_worker(
+	{ cwd, assets_base, config, manifest_data, build_dir, output_dir, service_worker_entry_file },
+	client_manifest
+) {
+	// TODO add any assets referenced in template .html file, e.g. favicon?
+	const app_files = new Set();
+	for (const key in client_manifest) {
+		const { file, css } = client_manifest[key];
+		app_files.add(file);
+		if (css) {
+			css.forEach((file) => {
+				app_files.add(file);
+			});
+		}
+	}
+
+	fs.writeFileSync(
+		`${build_dir}/runtime/service-worker.js`,
+		`
+			export const timestamp = ${Date.now()};
+
+			export const build = [
+				${Array.from(app_files)
+					.map((file) => `${s(`${config.kit.paths.base}/${config.kit.appDir}/${file}`)}`)
+					.join(',\n\t\t\t\t')}
+			];
+
+			export const files = [
+				${manifest_data.assets
+					.map((asset) => `${s(`${config.kit.paths.base}/${asset.file}`)}`)
+					.join(',\n\t\t\t\t')}
+			];
+		`
+			.replace(/^\t{3}/gm, '')
+			.trim()
+	);
+
+	/** @type {import('vite').UserConfig} */
+	const vite_config = config.kit.vite();
+
+	const default_config = {
```

Comment:
you can remove this (same in server and client). it'll be the default in Vite in a couple of days: https://vitejs.dev/config/#server-fs-strict